tutorials/developer-tips.md

```diff
@@ -0,0 +1,80 @@
+---
+title: Developer tips
+subtitle: 
+published: 
+code: 
+difficulty: 4
+---
+
+# Expected tokens #
+When you get a [`ParserError`](https://www.stackage.org/haddock/lts-7.18/megaparsec-5.0.1/Text-Megaparsec.html#t:ParseError), you can use the `errorExpected` function together with the [`toList`](https://www.stackage.org/haddock/lts-7.18/containers-0.5.7.1/Data-Set.html#v:toList) to get a list of expected [`ErrorItem`'s](https://www.stackage.org/haddock/lts-7.18/megaparsec-5.0.1/Text-Megaparsec.html#t:ErrorItem). These `ErrorItem`'s can be one of the following:
+
+* `Tokens`, in case there is a single token that was expected (such as a single character).
+* `Label`, in case a parser with a certain label comes next which expects various tokens.
+* `EndOfInput`, in case there are no more parsers.
+
+Expected here does not mean that, for example, the indicated token `"A"` MUST be matched. This is only true when there is a single item in the expected list. In case of multiple items only one of them has to be matched.
+
+# Developing your parser #
+
+## Top-down approach (don't do it) ##
+When you are re-using an existing grammar (or even when developing your own) you might be inclined to read the grammar top-down and also start developing your parser in a top-down fashion. There are various pitfalls when you develop your parser top-down. These pitfalls car occur when you want to test parts of your parser already even when it's not complete.
```

Comment:
It may be true that top-down approach is a dangerous path for beginners, I however succedded several times in developing parsers this way, so the phrase “don't do it” certainly is too strong.